src/xkbcomp/ast-build.c

```diff
@@ -200,6 +206,73 @@ ExprAppendKeysymList(ExprDef *expr, xkb_keysym_t sym)
     return expr;
 }
 
+ExprDef *
+ExprKeysymListAppendString(struct scanner *scanner,
+                           ExprDef *expr, const char *string)
+{
+    /* TODO: use strnlen with max len = 4 * MAX_KEYSYMS_LIST_LENGTH */
+    size_t len = strlen(string);
+    while (len) {
+        size_t count = 0;
+        uint32_t cp = utf8_next_code_point(string, len, &count);
+        if (cp == INVALID_UTF8_CODE_POINT) {
+            scanner_err(scanner, XKB_ERROR_INVALID_FILE_ENCODING,
```

Comment:
Using `scanner_err` from the parser is a little problematic because the location might be off (if the parse rule requires further tokens before the action is called). But I think it's not a blocker (we probably already have a couple of these).

Reply:
Interesting. But `STRING` is currently always the last token when it is called, isn’t it? I added indication of the failing string byte position + code point position, plus the corresponding log test.